Add input field for VATIN

lesser1 opened the pull request:
- Add input field to AddressInformationView.vue
- Add two new rules to check its input

[The corresponding Gropius Issue](https://frontend.gropius.duckdns.org/components/9d12d6ac-e85a-407c-bdff-a8321fb75d3a/issues/932f038d-8d1c-4259-a3fa-d51393915884)

### Definition of Done
- [x] Requirements of the issue are met
- [x] Code has been reviewed
- [x] Code is documented

lesser1 committed:
Add input field for VATIN

- Add input field to AddressInformationView.vue
- Add two new rules to check its input

nk-coding commented on src/views/checkout/AddressInformationView.vue:
```diff
@@ -47,6 +58,13 @@ watch(
     () => onSelectedDeliveryAddressChanged()
 )
 
+const _VATIN = ref<string | undefined>(order.value.vatNumber)
```
is this consistent with how you named other refs?
looks a bit strange to me

lesser1 replied:
No, it's not consistent. I considered `vatIdentificationNumber`, `vatNumber`, `VATIN` and `_VATIN`. I ended up with the last option because I wanted the ref to reflect the real world's acronym which is _VATIN_ but I did not want the ref to start with an uppercase letter. At the end of the day, it's only a ref. Consistency is a beautiful thing, so I will rename it to `vatNumber` -- also, to keep it simple.

lesser1 replied:
00fb73f28089e2bd0ced3fa889dac139e14d6745

nk-coding commented on src/util/rules.ts:
```diff
@@ -118,3 +118,43 @@ export function isValidCreditCardExpirationDate(
+ * @returns Returns true if the input is undefined, an empty string, or a valid VATIN. Otherwise, returns a string error message indicating the validation failure.
+ */
+export function isUndefinedOrEmptyStringOrValidVATIN(
+    input: string
```
the docs say this can be undefined, however the type says sth else

lesser1 replied:
Good point...

lesser1 replied:
00fb73f28089e2bd0ced3fa889dac139e14d6745

nk-coding commented on src/util/rules.ts:
```diff
@@ -118,3 +118,43 @@ export function isValidCreditCardExpirationDate(
+        return false
+    }
+
+    const patternToMatch = /^(EU|[A-Z]{2}|EL|XI)[\dA-Z]{2,13}$/
```
where does this come from? at least the Wikipedia article also mentions that dashes are used quite commonly, and I do not think this regex allows it

lesser1 replied:
Which article are you referring to? I could only find that Poland allows hyphens (dashes) for readability but formally their numbers only consist of digits. Of course, we could allow hyphens (dashes) and simply remove them as soon as we pass the value to the service. How about that?

lesser1 replied:
00fb73f28089e2bd0ced3fa889dac139e14d6745

nk-coding replied:
the article you reference in your docs: https://en.wikipedia.org/wiki/VAT_identification_number

below, there are some non-EU countries, some use hyphen
but I don't know if these are required

nk-coding commented on src/views/checkout/AddressInformationView.vue:
```diff
@@ -9,6 +9,16 @@
+                clearable
+                label="VAT Identification Number"
+                prepend-icon="mdi-card-account-details"
+                required
```
why is this required? the rule also seems to allow an empty string

lesser1 replied:
Copy-paste mistake... my bad. Will remove it.

lesser1 replied:
00fb73f28089e2bd0ced3fa889dac139e14d6745

nk-coding commented on src/views/checkout/AddressInformationView.vue:
```diff
@@ -47,6 +58,13 @@ watch(
+
+watch(
+    () => _VATIN.value,
+    () => updateVATINOfOrderInformation()
```
so, why not directly use order.value.vatNumber for the model of the input? or a computed ref at least?

lesser1 replied:
Since I do not have a convincing reason to keep the ref in the middle, I will change it. This then also resolves the issue regarding the naming of the `_VATIN` ref.

lesser1 replied:
00fb73f28089e2bd0ced3fa889dac139e14d6745

lesser1 committed:
Modify handling of VATIN

- Remove unnecessary ref
- Remove incorrect required attribute from input field
- Modify regular expression for VATIN to allow for hyphens
- Add string utility function that removes all hyphens from a string
- Add workaround at order creation to turn undefined VATIN into an
  empty string
